packages/vm/src/runTx.ts

```diff
@@ -320,7 +322,24 @@ async function _runTx(this: VM, opts: RunTxOpts): Promise<RunTxResult> {
   /*
    * Execute message
    */
-  const { value, data, to } = tx
+  const { value, to } = tx
+  let { data } = tx
+
+  // Sapphire decode call data
+  let format, aead
+  if (this.confidential && data !== null && data.length > 0) {
+    const { format, body: envelopBody } = cbor.decode(data)
+    if (format === 1) {
+      // X25519DeoxysII
+      const { nonce: deoxysiiNonce, data: envelopeData, pk } = envelopBody
+      aead = cipher.X25519DeoxysII.fromSecretKey(this.secretKey, pk)
+      const { body } = await aead.decryptCallData(deoxysiiNonce, envelopeData)
+      data = Buffer.from(body.buffer)
```

Comment:
```suggestion
      data = Buffer.from(body)
```
and similarly below. If you `from` on just the `buffer` and ignore the `byteOffset` and `length`, you will get a very confusing error message if the typed array returned by `decryptCallData` is a slice of a larger allocated chunk.